core/src/main/scala/com/databricks/labs/remorph/coverage/estimation/Estimator.scala

```diff
@@ -0,0 +1,182 @@
+                      EstimateAnalysisReport(Some(fingerprint)))
+                }
+              } else {
+                // Empty report will be filtered out as it means it was a duplicate
+                EstimateReportRecord(EstimateTranspilationReport(), EstimateAnalysisReport())
+              }
+            case _ =>
+              EstimateReportRecord(
+                EstimateTranspilationReport(
+                  query = Some(query.source),
+                  statements = 1,
+                  parsing_error = Some("Unexpected result from parse phase")),
+                EstimateAnalysisReport())
+          }
+        }
+        .filter(_.transpilationReport.statements > 0)
+
+      val (uniqueSuccesses, parseFailures, transpileFailures) = reportEntries.foldLeft((0, 0, 0)) {
+        case ((uniqueSuccesses, parseFailures, transpileFailures), entry) =>
+          val newUniqueSuccesses =
+            if (entry.transpilationReport.parsed == 1 && entry.transpilationReport.transpiled == 1) uniqueSuccesses + 1
+            else uniqueSuccesses
+          val newParseFailures =
+            if (entry.transpilationReport.parsing_error.isDefined) parseFailures + 1 else parseFailures
+          val newTranspileFailures =
+            if (entry.transpilationReport.transpilation_error.isDefined) transpileFailures + 1 else transpileFailures
+          (newUniqueSuccesses, newParseFailures, newTranspileFailures)
+      }
+
+      val report = EstimateReport(
+        sampleSize = history.queries.size,
+        uniqueSuccesses = uniqueSuccesses,
+        parseFailures = parseFailures,
+        transpileFailures = transpileFailures,
+        records = reportEntries,
+        overallComplexity = SqlComplexity.LOW // Will calculate this later
+      )
+
+      val jsonReport: String = write(report, indent = 4)
+      os.write(resultPath, jsonReport)
+
+      if (consoleReport) {
+        // scalastyle:off println
+        println(s"Sample size        : ${history.queries.size}")
+        println(s"Unique successes   : $uniqueSuccesses")
+        println(s"Parse failures     : $parseFailures")
+        println(s"Transpile failures : $transpileFailures")
+        println(s"Parsed query count : ${report.uniqueSuccesses}")
+        println(s"Overall complexity : ${report.overallComplexity}")
+        println(s"Report written to  : $resultPath")
+        println()
+        // scalastyle:on println
+      }
+    } finally {
+      conn.close()
+    }
+  }
+
+  // TODO: Probably needs to go into a plan analyzer class, along with SqlComplexity
+  private def countStatements(plan: ir.TreeNode[_]): Int = {
+    try {
+      plan
+        .collect {
+          case b: ir.Batch if b.output != null =>
+            b.output.size
+          case b: ir.Batch =>
+            logger.warn(s"Unexpected batch with no children: $b")
+            0
+          case _: ir.ScalarSubquery =>
+            1
+          case _ => 0
+        }
```

Comment:
```suggestion
          case b: ir.Batch if b.output != null => b.output.size
          case b: ir.Batch =>
            logger.warn(s"Unexpected batch with no children: $b")
            0
          case _: ir.ScalarSubquery => 1
          case _ => 0
        }
```

i'd even separate a warning to a separate expression, so that it's more readable:

```suggestion
          case b: ir.Batch if b.output != null => b.output.size
          case b: ir.Batch => 0
          case _: ir.ScalarSubquery => 1
          case _ => 0
        }
```